spec/classes/init_spec.rb

```diff
@@ -1,395 +1,141 @@
+    let(:params) { { :package_ensure => 'absent' } }
 
-      it {
-        should contain_service('rpcbind_service').with({
-          'ensure' => 'running',
-          'name'   => 'rpcbind',
-          'enable' => true,
-        })
-      }
-    end
+    it {
+      should contain_package('rpcbind_package').with({
+        'ensure' => 'absent',
+      })
+    }
+  end
 
-    context 'with default params on Suse 12' do
-      let(:facts) do
-        { :osfamily          => 'Suse',
-          :lsbmajdistrelease => '12',
-        }
-      end
+  describe 'with package_name specified' do
+    let(:facts) { platforms['el7'][:facts_hash] }
+    let(:params) { { :package_name => 'my_rpcbind' } }
 
-      it {
-        should contain_service('rpcbind_service').with({
-          'ensure' => 'running',
-          'name'   => 'rpcbind',
-          'enable' => true,
-        })
-      }
-    end
+    it {
+      should contain_package('rpcbind_package').with({
+        'ensure' => 'installed',
+        'name'   => 'my_rpcbind',
+      })
+    }
+  end
 
-    context 'with ensure stopped' do
-      let(:facts) { { :osfamily => 'RedHat' } }
-      let(:params) { { :service_ensure => 'stopped' } }
+  describe 'with service_ensure stopped' do
+    let(:facts) { platforms['el7'][:facts_hash] }
+    let(:params) { { :service_ensure => 'stopped' } }
 
-      it {
-        should contain_service('rpcbind_service').with({
-          'ensure' => 'stopped',
-          'name'   => 'rpcbind',
-          'enable' => true,
-        })
-      }
-    end
+    it {
+      should contain_service('rpcbind_service').with({
+        'ensure' => 'stopped',
+      })
+    }
+  end
 
-    context 'with supplied string for service name' do
-      let(:facts) { { :osfamily => 'RedHat' } }
-      let(:params) { { :service_name => 'my_rpcbind' } }
+  describe 'with service_name specified' do
+    let(:facts) { platforms['el7'][:facts_hash] }
+    let(:params) { { :service_name => 'my_rpcbind' } }
 
-      it {
-        should contain_service('rpcbind_service').with({
-          'ensure' => 'running',
-          'name'   => 'my_rpcbind',
-          'enable' => true,
-        })
-      }
-    end
+    it {
+      should contain_service('rpcbind_service').with({
+        'name' => 'my_rpcbind',
+      })
+    }
+  end
 
-    context 'with enable false' do
-      let(:facts) { { :osfamily => 'RedHat' } }
-      let(:params) { { :service_enable => false } }
+  describe 'with service_enable false' do
+    let(:facts) { platforms['el7'][:facts_hash] }
+    let(:params) { { :service_enable => false } }
 
-      it {
-        should contain_service('rpcbind_service').with({
-          'ensure' => 'running',
-          'name'   => 'rpcbind',
-          'enable' => false,
-        })
-      }
-    end
+    it {
+      should contain_service('rpcbind_service').with({
+        'enable' => false,
+      })
+    }
   end
+
+  mandatory_facts = platforms['el7'][:facts_hash]
```

Comment:
@Phil-Friderici I get this error with or without specifying this line. Any ideas why it is always failing on the valid matches? Seems that it doesn't know the `$facts['os']['family']` though I thought I am specifying it.

The error is

```
     Failure/Error: it { should compile }
       error during compilation: Evaluation Error: Operator '[]' is not applicable to an Undef Value. (file: /Users/gh/git/puppet-module-rpcbind/spec/fixtures/modules/rpcbind/manifests/init.pp, line: 23, column: 6)
```

Reply:
Can't test myself at the moment. Could you please try to add curly brackets like this:

```mandatory_facts = { platforms['el7'][:facts_hash] }```

Reply:
@Phil-Friderici Thanks for having a look!

That gives an error.

added some puts to give the following output with the code as it currently is.

```
mandatory_facts class is Hash and the value is

{
  :osfamily=>"RedHat",
  :operatingsystem=>"RedHat",
  :operatingsystemmajrelease=>"7",
  :os=> {
    "name"=>"RedHat",
    "family"=>"RedHat",
    "release"=> {
      "full"=>"7.0.1406",
      "major"=>"7",
      "minor"=>"0"
    }
  }
}
```

Made some other changes, though still getting the same error. If I comment out the platform validation parts in the manifest, then the type validation passes, so it seems that it does not know about the specified facts.

Added another commit.

Reply:
In general there is no need to explicilty set the variable ```mandatory_facts```. It can be empty.
I had added and used it to set non OS related facts in the past.

Guess it will be simplier to just set the OS like in the other cases with:
```
describe 'variable data type and content validations' do
  let(:facts) { platforms['el7'][:facts_hash] }

  validations = {
```

Reply:
great idea! that worked!